File: backend-nestjs/src/controllers/admin-sitemap.controller.ts
import { Body, Controller, Delete, Get, Param, Post, Put, Query, Req } from '@nestjs/common';
import type { Request } from 'express';
import { AdminAuthorizationService } from '../auth/admin-authorization.service';
import { ServicePermission } from '../auth/admin-permissions';
import { SitemapService } from '../services/sitemap.service';
import { SitemapSettingsUpsertRequestDto } from '../dto/requests/sitemap-settings-upsert.dto';
import { SitemapSettingsResponseDto } from '../dto/responses/sitemap-settings-response.dto';
import { SitemapTemplateResponseDto } from '../dto/responses/sitemap-template-response.dto';
import { SitemapTemplateUpsertRequestDto } from '../dto/requests/sitemap-template-upsert.dto';
import { SitemapPreviewEntryResponseDto } from '../dto/responses/sitemap-preview-entry-response.dto';
import { SitemapTestUrlRequestDto } from '../dto/requests/sitemap-test-url.dto';
import { SitemapTestUrlResponseDto } from '../dto/responses/sitemap-test-url-response.dto';
import { SitemapOverrideUpsertRequestDto } from '../dto/requests/sitemap-override-upsert.dto';
import { SitemapCustomUrlResponseDto } from '../dto/responses/sitemap-custom-url-response.dto';
import { SitemapCustomUrlUpsertRequestDto } from '../dto/requests/sitemap-custom-url-upsert.dto';

@Controller('/api/v1/admin/sitemap')
export class AdminSitemapController {
  constructor(
    private readonly sitemapService: SitemapService,
    private readonly access: AdminAuthorizationService,
  ) {}

  private assertAccess(request: Request, applicationId: string): void {
    this.access.assertAnyServicePermission(request, [
      ServicePermission.POSTS_MANAGE,
      ServicePermission.ARTICLES_MANAGE,
      ServicePermission.IMAGES_MANAGE,
      ServicePermission.VIDEOS_MANAGE,
      ServicePermission.COLLECTIONS_MANAGE,
    ]);
    this.access.assertApplicationAccess(request, applicationId);
  }

  @Get('settings')
  async getSettings(
    @Req() request: Request,
    @Query('applicationId') applicationId: string,
  ): Promise<SitemapSettingsResponseDto> {
    this.assertAccess(request, applicationId);
    return await this.sitemapService.getSettings(applicationId);
  }

  @Put('settings')
  async putSettings(
    @Req() request: Request,
    @Query('applicationId') applicationId: string,
    @Body() body: SitemapSettingsUpsertRequestDto,
  ): Promise<SitemapSettingsResponseDto> {
    this.assertAccess(request, applicationId);
    return await this.sitemapService.putSettings(applicationId, body);
  }

  @Get('templates')
  async listTemplates(
    @Req() request: Request,
    @Query('applicationId') applicationId: string,
  ): Promise<SitemapTemplateResponseDto[]> {
    this.assertAccess(request, applicationId);
    return await this.sitemapService.listTemplates(applicationId);
  }

  @Put('templates/:contentType')
  async putTemplate(
    @Req() request: Request,
    @Query('applicationId') applicationId: string,
    @Param('contentType') contentType: string,
    @Body() body: SitemapTemplateUpsertRequestDto,
  ): Promise<SitemapTemplateResponseDto> {
    this.assertAccess(request, applicationId);
    return await this.sitemapService.putTemplate(applicationId, contentType, body);
  }

  @Get('preview')
  async preview(
    @Req() request: Request,
    @Query('applicationId') applicationId: string,
    @Query('contentType') contentType?: string,
    @Query('limit') limit = '50',
    @Query('offset') offset = '0',
  ): Promise<{ total: number; items: SitemapPreviewEntryResponseDto[] }> {
    this.assertAccess(request, applicationId);
    return await this.sitemapService.preview(applicationId, contentType, Number(limit), Number(offset));
  }

  @Post('test-url')
  async testUrl(
    @Req() request: Request,
    @Query('applicationId') applicationId: string,
    @Body() body: SitemapTestUrlRequestDto,
  ): Promise<SitemapTestUrlResponseDto> {
    this.assertAccess(request, applicationId);
    return await this.sitemapService.testUrl(applicationId, body.url);
  }

  @Put('override/:contentType/:contentId')
  async putOverride(
    @Req() request: Request,
    @Query('applicationId') applicationId: string,
    @Param('contentType') contentType: string,
    @Param('contentId') contentId: string,
    @Body() body: SitemapOverrideUpsertRequestDto,
  ): Promise<{ ok: true }> {
    this.assertAccess(request, applicationId);
    await this.sitemapService.putOverride(applicationId, contentType, contentId, body);
    return { ok: true };
  }

  @Get('custom-urls')
  async listCustomUrls(
    @Req() request: Request,
    @Query('applicationId') applicationId: string,
  ): Promise<SitemapCustomUrlResponseDto[]> {
    this.assertAccess(request, applicationId);
    return await this.sitemapService.listCustomUrls(applicationId);
  }

  @Post('custom-urls')
  async createCustomUrl(
    @Req() request: Request,
    @Query('applicationId') applicationId: string,
    @Body() body: SitemapCustomUrlUpsertRequestDto,
  ): Promise<SitemapCustomUrlResponseDto> {
    this.assertAccess(request, applicationId);
    return await this.sitemapService.createCustomUrl(applicationId, body);
  }

  @Put('custom-urls/:id')
  async updateCustomUrl(
    @Req() request: Request,
    @Query('applicationId') applicationId: string,
    @Param('id') id: string,
    @Body() body: SitemapCustomUrlUpsertRequestDto,
  ): Promise<SitemapCustomUrlResponseDto> {
    this.assertAccess(request, applicationId);
    return await this.sitemapService.updateCustomUrl(applicationId, id, body);
  }

  @Delete('custom-urls/:id')
  async deleteCustomUrl(
    @Req() request: Request,
    @Query('applicationId') applicationId: string,
    @Param('id') id: string,
  ): Promise<{ ok: true }> {
    this.assertAccess(request, applicationId);
    await this.sitemapService.deleteCustomUrl(applicationId, id);
    return { ok: true };
  }
}

File: backend/src/main/java/com/contentplatform/backend/interfaces/web/controller/AdminSitemapController.java
package com.contentplatform.backend.interfaces.web.controller;

import com.contentplatform.backend.application.exception.ForbiddenException;
import com.contentplatform.backend.application.service.SitemapService;
import com.contentplatform.backend.domain.value.ServicePermission;
import com.contentplatform.backend.interfaces.web.SecurityUtils;
import com.contentplatform.backend.interfaces.web.request.SitemapCustomUrlRequest;
import com.contentplatform.backend.interfaces.web.request.SitemapOverrideRequest;
import com.contentplatform.backend.interfaces.web.request.SitemapSettingsRequest;
import com.contentplatform.backend.interfaces.web.request.SitemapTemplateRequest;
import com.contentplatform.backend.interfaces.web.request.SitemapTestUrlRequest;
import com.contentplatform.backend.interfaces.web.response.SitemapCustomUrlResponse;
import com.contentplatform.backend.interfaces.web.response.SitemapPreviewResponse;
import com.contentplatform.backend.interfaces.web.response.SitemapSettingsResponse;
import com.contentplatform.backend.interfaces.web.response.SitemapTemplateResponse;
import com.contentplatform.backend.interfaces.web.response.SitemapTestUrlResponse;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import java.util.List;
import java.util.Set;

@RestController
@RequestMapping("/api/v1/admin/sitemap")
public class AdminSitemapController {
    private static final Set<ServicePermission> ALLOWED = Set.of(
        ServicePermission.POSTS_MANAGE,
        ServicePermission.ARTICLES_MANAGE,
        ServicePermission.VIDEOS_MANAGE,
        ServicePermission.COLLECTIONS_MANAGE
    );

    private final SitemapService sitemapService;

    public AdminSitemapController(SitemapService sitemapService) {
        this.sitemapService = sitemapService;
    }

    @GetMapping("/settings")
    public ResponseEntity<SitemapSettingsResponse> getSettings(@RequestParam String applicationId) {
        assertAccess(applicationId);
        return ResponseEntity.ok(sitemapService.getSettings(applicationId));
    }

    @PutMapping("/settings")
    public ResponseEntity<SitemapSettingsResponse> putSettings(@RequestParam String applicationId,
                                                               @RequestBody SitemapSettingsRequest request) {
        assertAccess(applicationId);
        return ResponseEntity.ok(sitemapService.updateSettings(applicationId, request));
    }

    @GetMapping("/templates")
    public ResponseEntity<List<SitemapTemplateResponse>> listTemplates(@RequestParam String applicationId) {
        assertAccess(applicationId);
        return ResponseEntity.ok(sitemapService.listTemplates(applicationId));
    }

    @PutMapping("/templates/{contentType}")
    public ResponseEntity<SitemapTemplateResponse> putTemplate(@RequestParam String applicationId,
                                                               @PathVariable String contentType,
                                                               @RequestBody SitemapTemplateRequest request) {
        assertAccess(applicationId);
        return ResponseEntity.ok(sitemapService.upsertTemplate(applicationId, contentType, request));
    }

    @GetMapping("/preview")
    public ResponseEntity<SitemapPreviewResponse> preview(@RequestParam String applicationId,
                                                          @RequestParam(required = false) String contentType,
                                                          @RequestParam(defaultValue = "50") int limit,
                                                          @RequestParam(defaultValue = "0") int offset) {
        assertAccess(applicationId);
        return ResponseEntity.ok(sitemapService.preview(applicationId, contentType, limit, offset));
    }

    @PostMapping("/test-url")
    public ResponseEntity<SitemapTestUrlResponse> testUrl(@RequestParam String applicationId,
                                                          @RequestBody SitemapTestUrlRequest request) {
        assertAccess(applicationId);
        return ResponseEntity.ok(sitemapService.testUrl(applicationId, request.getUrl()));
    }

    @PutMapping("/override/{contentType}/{contentId}")
    public ResponseEntity<Void> putOverride(@RequestParam String applicationId,
                                            @PathVariable String contentType,
                                            @PathVariable String contentId,
                                            @RequestBody SitemapOverrideRequest request) {
        assertAccess(applicationId);
        sitemapService.upsertOverride(applicationId, contentType, contentId, request);
        return ResponseEntity.ok().build();
    }

    @GetMapping("/custom-urls")
    public ResponseEntity<List<SitemapCustomUrlResponse>> listCustomUrls(@RequestParam String applicationId) {
        assertAccess(applicationId);
        return ResponseEntity.ok(sitemapService.listCustomUrls(applicationId));
    }

    @PostMapping("/custom-urls")
    public ResponseEntity<SitemapCustomUrlResponse> createCustomUrl(@RequestParam String applicationId,
                                                                    @RequestBody SitemapCustomUrlRequest request) {
        assertAccess(applicationId);
        return ResponseEntity.ok(sitemapService.createCustomUrl(applicationId, request));
    }

    @PutMapping("/custom-urls/{id}")
    public ResponseEntity<SitemapCustomUrlResponse> updateCustomUrl(@RequestParam String applicationId,
                                                                    @PathVariable String id,
                                                                    @RequestBody SitemapCustomUrlRequest request) {
        assertAccess(applicationId);
        return ResponseEntity.ok(sitemapService.updateCustomUrl(applicationId, id, request));
    }

    @DeleteMapping("/custom-urls/{id}")
    public ResponseEntity<Void> deleteCustomUrl(@RequestParam String applicationId, @PathVariable String id) {
        assertAccess(applicationId);
        sitemapService.deleteCustomUrl(applicationId, id);
        return ResponseEntity.noContent().build();
    }

    private void assertAccess(String applicationId) {
        SecurityUtils.requireApplicationAccess(applicationId);
        List<ServicePermission> permissions = SecurityUtils.getServicePermissions();
        if (permissions.isEmpty()) {
            return;
        }
        boolean allowed = permissions.stream().anyMatch(ALLOWED::contains);
        if (!allowed) {
            throw new ForbiddenException("Service permission denied");
        }
    }
}

